From 444e3d375165082595b7d75d8e8206acafd2a205 Mon Sep 17 00:00:00 2001
From: Thomas Preud'homme <robotux@celest.fr>
Date: Thu, 15 Mar 2012 00:22:00 +0100
Subject: Support linker options passed in several -Wl param

ld support arguments to multiple-letter options being passed in two
ways:
* -opt=arg
* -opt arg

libtool generate command line of the second form. This commit add
support for the second form so that libtool works with tcc. The way it
is done is to concatenate all -Wl options into one and then pass it to
set_linker.

Origin: vendor
Bug-Debian: http://bugs.debian.org/663693
Forwarded: http://lists.nongnu.org/archive/html/tinycc-devel/2012-03/msg00033.html
Last-Update: 2012-03-20
Applied-Upstream: commit: 1c11b857fe87a73d7047701dacdbe8394967b1d1
---
 libtcc.c |    3 +++
 tcc.c    |   13 +++++++++++--
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/libtcc.c b/libtcc.c
index 8f5cd0e..8aadc8f 100644
--- a/libtcc.c
+++ b/libtcc.c
@@ -1566,6 +1566,9 @@ PUB_FUNC const char * tcc_set_linker(TCCState *s, char *option, int multi)
             s->has_text_addr = 1;
 
         } else {
+            char *comma_ptr = strchr(option, ',');
+            if (comma_ptr)
+                *comma_ptr = '\0';
             return option;
         }
 
diff --git a/tcc.c b/tcc.c
index 0c51451..aadd273 100644
--- a/tcc.c
+++ b/tcc.c
@@ -35,6 +35,7 @@ static int do_bench = 0;
 static int gen_deps;
 static const char *deps_outfile;
 static const char *m_option;
+static CString linker_arg;
 
 #define TCC_OPTION_HAS_ARG 0x0001
 #define TCC_OPTION_NOSEP   0x0002 /* cannot have space before option and arg */
@@ -284,6 +285,7 @@ static int parse_args(TCCState *s, int argc, char **argv)
     was_pthread = 0; /* is set if commandline contains -pthread key */
 
     optind = 1;
+    cstr_new(&linker_arg);
     while (optind < argc) {
 
         r = argv[optind++];
@@ -442,8 +444,12 @@ static int parse_args(TCCState *s, int argc, char **argv)
                 s->rdynamic = 1;
                 break;
             case TCC_OPTION_Wl:
-                if ((r = (char *) tcc_set_linker(s, (char *)optarg, TRUE)))
-                    tcc_error("unsupported linker option '%s'", r);
+                if (!linker_arg.data_allocated)
+                    cstr_cat(&linker_arg, optarg);
+                else {
+                    cstr_ccat(&linker_arg, ',');
+                    cstr_cat(&linker_arg, optarg);
+                }
                 break;
             case TCC_OPTION_E:
                 output_type = TCC_OUTPUT_PREPROCESS;
@@ -465,6 +471,8 @@ static int parse_args(TCCState *s, int argc, char **argv)
             }
         }
     }
+    if ((r = (char *) tcc_set_linker(s, (char *) linker_arg.data, TRUE)))
+        tcc_error("unsupported linker option '%s'", r);
     /* fixme: these options could be different on your platform */
     if (was_pthread && output_type != TCC_OUTPUT_OBJ) {
         dynarray_add((void ***)&files, &nb_files, "-lpthread");
@@ -593,6 +601,7 @@ int main(int argc, char **argv)
     }
 
     tcc_delete(s);
+    cstr_free(&linker_arg);
     tcc_free(outfile);
 
 #ifdef MEM_DEBUG
